R/best_case_worst_case.R

```diff
@@ -0,0 +1,114 @@
+#' Calculates Best case and Worst case for Emission profile at product level
+#'
+#' @param data Dataframe. Emissions profile product level output
+#'
+#' @return A dataframe
+#' @export
+#'
+#' @examples
+#' library(tiltToyData)
+#' library(readr)
+#' restore <- options(readr.show_col_types = FALSE)
+#'
+#' companies <- read_csv(toy_emissions_profile_any_companies())
+#' co2 <- read_csv(toy_emissions_profile_products_ecoinvent())
+#' europages_companies <- read_csv(toy_europages_companies())
+#' ecoinvent_activities <- read_csv(toy_ecoinvent_activities())
+#' ecoinvent_europages <- read_csv(toy_ecoinvent_europages())
+#' isic_name <- read_csv(toy_isic_name())
+#' profile_emissions_at_product_level <- profile_emissions(
+#'   companies,
+#'   co2,
+#'   europages_companies = europages_companies,
+#'   ecoinvent_activities = ecoinvent_activities,
+#'   ecoinvent_europages = ecoinvent_europages,
+#'   isic = isic_name
+#' ) |>
+#'   unnest_product()
+#'
+#' result <- best_case_worst_case(profile_emissions_at_product_level)
```

Comment:
@maurolepore Instead, can I insert the `bese_case_worst_case()` function inside the `polish_emissions_any()` function? I am asking this because `bese_case_worst_case` can only be applied to the product level output of profile emissions. And, you are using the `unnest_product()` inside the `polish_emissions_any()` function. So, the `polish_emissions_any()` function will look like this:

```
polish_emissions_any <- function(data,
                                 ...,
                                 output_min_max = FALSE,
                                 output_co2_footprint = FALSE) {
  product <- data |>
    unnest_product() |>
    polish_emissions_any_impl(
      output_min_max = output_min_max,
      output_co2_footprint = output_co2_footprint
    ) |>
    best_case_worst_case()

  company <- data |>
    unnest_company()

  tilt_profile(nest_levels(product, company))
}
```

Reply:
@kalashsinghal,

The fact that `polish_*()` calls `unnest_product()` ([here](    unnest_product() |>)) doesn't force you at all to move `best_case_worst_case()` inside `polish_*()`. In fact it also calls `unnest_company()` ([here](https://github.com/2DegreesInvesting/tiltIndicatorAfter/blob/acdf2a124643b40e675c4a7bf650a5a27c68bf6c/R/polish_emissions_any.R#L13)). 

`polish_*()` takes the entire [tilt_profile object](https://2degreesinvesting.github.io/tiltIndicator/reference/tilt_profile) so that it can be part of the pipe. If you do the same with `best_case_worst_case()` then it can also be part of the pipe:

```r
profile_emissions_impl() |>
  add_co2() |>
  bese_case_worst_case() |>
  polish_emissions_any()
```

To enable this modular design every function ("mudule") takes the entire tilt_profile object as the first input -- irrespective of whether they internally use `product`, `company`, both or none. The standard connections (the tilt_profile object) enables modular composition.

You'll find fancy names for the pattern (like "ports and adapters") but the principle is simple and common in everyday life. Think of how much easier it would be to travel if every country used the same type of electric port. You could connect your electric devices anywhere without an adepter :-)

### But maybe some code needs to move to `polish_*()`?

In general the `polish_*()` functions should do no computations. It should only adapt the data from the form that is most convenient for the internal modules to the for that Anne requests. In most cases this means that `polish_*()` would only rename columns. 

In terms of the clean architecture model, `polish_*()` is an interface adaptor. It's a pretty low-level detail, laying at close to the edge -- near the ouput.

https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html

![image](https://github.com/2DegreesInvesting/tiltIndicatorAfter/assets/5856545/4145b1e6-cd01-4ad3-b7f3-6636ee379fad)


Therefore, if some code in `best_case_worst_case()` deals with responsabilities of `polish_*()` then that piece of code should indeed move to `polish_*()`. But everything else should stay in more central layers of the architecture.

Reply:
@maurolepore Thanks for this info! Very helpful :)

This [commit](https://github.com/2DegreesInvesting/tiltIndicatorAfter/pull/243/commits/6c3ced415f51438b7e0b445a69e2af3ab6381253) tries to adapt the "clean architecture". Please let me know of any changes required!

Reply:
@maurolepore I have created a new wrapper with name `best_case_worst_case` which wraps the old function `best_case_worst_case_emission_profile` (renamed). Both are used in this way:

```
profile_emissions_impl() |>
  add_co2() |>
  bese_case_worst_case(best_case_worst_case_emission_profile) |>
  polish_emissions_any()
```

Here `best_case_worst_case` does the nesting of product and company level results. This wrapper can also be used for applying best and worst case to other indicators in the future.